File: medium/53. Maximum Subarray/53. Maximum Subarray.ts
function maxSubArray(nums: number[]): number {
    let res = -Infinity
    let currSum = 0
    const n = nums.length

    for (let i = 0; i < n; i++) {
        if (currSum < 0) {
            currSum = 0
        }

        currSum = currSum + nums[i]
        res = Math.max(currSum, res)
    }

    return res
};

File: medium/53. Maximum Subarray/53. Maximum Subarray.py
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # Глобальный результат подсчета
        res = nums[0]

        # Максимальная сумма элементов до тех пор, пока не будет найден наибольший i элемент
        max_ending = nums[0]
        n = len(nums)

        for i in range(1, n):
            max_ending = max(max_ending + nums[i], nums[i])
            res = max(res, max_ending)

        return res
